rfcs/001 - EC2 schema.md

```diff
@@ -0,0 +1,136 @@
+- Launch templates
+- Spot instances
+- Dedicated hosts
+- Saving plans
+- Reserved instances
+- Scheduled instances
+- Capacity reservations
+- Auto Scaling Groups
+- AMIs management
+- EBS volume snapshots and life cycle
+- Placement groups
+- Key Pairs
+- Network interfaces
+
+The `instance` table still needs more advanced options to be defined. Some of those missing properties are: instance auto-recovery, shutdown behaviour, stop - hibernate behaviour, termination protection, stop protection, detailed cloudWatch monitoring, elastic GPU, credit specification, tenancy, RAM disk ID, kernel ID or Metadata accessible (I might be missing others). Additionally, some of the features listed above will need to be added to the `instance` table as FKs.
+
+The current EC2 model looks like this:
+
+![Current EC2 model](./assets/001/current_schema.png)
+Note: Could be also accessed via https://dbdocs.io/alejandro/iasql
+
+## Proposal
+
+After considering alternatives like splitting everything into more modules or having different `instance` tables around based on what service creates them, the proposal is to keep the schema growing in this same module. The relationships, though, will be added where they belong and rely on the instance metadata for the no needed ones for the CRUD of the `instance` table.
+
+We will review the needed features, see their implications in the model and get a final version of the ECS schema.
+
+- Creating an instance using the UI/API lets you add some advanced configuration that usually is not used but is there. The screenshot below shows part of the extra configuration. At least the following columns will need to be added to the `instance` table: instance auto-recovery, shutdown behaviour, stop - hibernate behaviour, termination protection, stop protection, detailed cloudWatch monitoring, elastic GPU, credit specification, tenancy, RAM disk ID, kernel ID or Metadata accessible.
+
+![EC2 instance advance properties](./assets/001/ec2_advance_properties.png)
+
+- Instance types depend on availability zones. We need to create a table relating them using a composite key. We also need to add the `availability_zone` column to the `instance` table to relate with the new `instance_type` table using the composite key.
+
+- Launch templates will be a new table and need to be an `instance` table input column `launch_template_id`. If we insert other values to the `instance`, these will override the launch template ones. To know if an instance was created using one of them we check the tags and look for the id to link the FK.
+
+- Spot requests can be created manually, using launch templates or can be created by other entities like ASG or the same instance. If an instance was created using a spot request should not affect the `instance` table, so the relationship between them will be part of the `instance_metadata`. To be able to know if an instance was created by a spot request we check the tags and look for the id to link them.
```

Comment:
The phrasing here is a bit ambiguous. Does "If an instance was created using a spot request should not affect the `instance` table mean:

1. That spot instances *don't* show up in the `instance` table and therefore do not affect it.
2. Spot instances *do* show up in the `instance` table, but do not affect the schema of the `instance` table.

Reply:
BTW, I am assuming it's (2) here, but I first read it as (1).

Reply:
Yes, it is the second. They do show up but if they were created or not by a spot request will not affect the schema of the instance table in terms of CRUD operations. Maybe this rephrase explain it better?

```suggestion
- Spot requests can be created manually, using launch templates or can be created by other entities like ASG or the same instance. The relationship between the `instance` and the spot request will be part of the `instance_metadata` since the spot request does not affect the `instance` schema in terms of CRUD operations. To be able to know if an instance was created by a spot request we check the tags and look for the id to link them.
```

Reply:
That rewrite clears it up!